gravitee-gateway-policy/src/main/java/io/gravitee/gateway/policy/impl/CachedPolicyFactory.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.gateway.policy.impl;
+
+import io.gravitee.gateway.policy.*;
+import io.gravitee.policy.api.PolicyConfiguration;
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * @author Jeoffrey HAEYAERT (jeoffrey.haeyaert at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class CachedPolicyFactory implements PolicyFactory {
+
+    private final ConcurrentMap<String, Policy> policies = new ConcurrentHashMap<>();
+    private final PolicyFactory delegate;
+
+    public CachedPolicyFactory(PolicyFactory delegate) {
+        Objects.requireNonNull(delegate, "PolicyFactory delegate is mandatory");
+
+        this.delegate = delegate;
+    }
+
+    @Override
+    public Policy create(StreamType streamType, PolicyMetadata policyMetadata, PolicyConfiguration policyConfiguration) {
+        final String key =
+            streamType.hashCode() +
+            "-" +
+            policyMetadata.hashCode() +
+            "-" +
+            (policyConfiguration == null ? "#null" : policyConfiguration.hashCode());
+        return policies.computeIfAbsent(key, k -> delegate.create(streamType, policyMetadata, policyConfiguration));
+    }
+
+    public void clear() {
+        policies.clear();
```

Comment:
@jhaeyaert When this method is called ?

Reply:
It is useless in our case as the factory is destroyed and recreated on each api deployment.
I was just thinking about the Platform policies where it could be useful but I didn't covered the case because I think there is work to do on it (ie: refactor).